File: backend/middleware/rateLimit.js
const rateLimit = require('express-rate-limit');
const { pool } = require('../db');
// Atomic counters shared by every application process, using the existing database.
class PostgresStore {
  constructor(prefix) { this.prefix = prefix; }
  init(options) { this.windowMs = options.windowMs; }
  async increment(key) {
    const { rows } = await pool.query(`INSERT INTO rate_limit_counters (key, hits, expires_at)
      VALUES ($1, 1, NOW() + $2 * INTERVAL '1 millisecond')
      ON CONFLICT (key) DO UPDATE SET
      hits = CASE WHEN rate_limit_counters.expires_at <= NOW() THEN 1 ELSE rate_limit_counters.hits + 1 END,
      expires_at = CASE WHEN rate_limit_counters.expires_at <= NOW() THEN EXCLUDED.expires_at ELSE rate_limit_counters.expires_at END
      RETURNING hits, expires_at`, [this.prefix + ':' + key, this.windowMs]);
    return { totalHits: rows[0].hits, resetTime: rows[0].expires_at };
  }
  async decrement(key) { await pool.query('UPDATE rate_limit_counters SET hits = GREATEST(0, hits - 1) WHERE key = $1', [this.prefix + ':' + key]); }
  async resetKey(key) { await pool.query('DELETE FROM rate_limit_counters WHERE key = $1', [this.prefix + ':' + key]); }
}
function sharedRateLimit(prefix, options) {
  return rateLimit({ ...options, store: new PostgresStore(prefix) });
}
module.exports = { sharedRateLimit, PostgresStore };

File: django_backend/apps/core/services/rate_limit.py
from django.db import connection
from django.utils import timezone
from datetime import timedelta

class PostgresRateLimiter:
    """
    Atomic PostgreSQL rate limiter matching Express sharedRateLimit.
    Uses the 'rate_limit_counters' table to share limits across processes.
    """
    def __init__(self, prefix: str, window_ms: int, max_hits: int):
        self.prefix = prefix
        self.window_ms = window_ms
        self.max_hits = max_hits

    def check_and_increment(self, key: str) -> tuple[bool, int, timezone.datetime]:
        """
        Increments the hit counter and checks if it exceeds max_hits.
        Returns: (is_allowed, current_hits, reset_time)
        """
        full_key = f"{self.prefix}:{key}"
        window_delta = timedelta(milliseconds=self.window_ms)
        now = timezone.now()
        expiry = now + window_delta

        with connection.cursor() as cursor:
            cursor.execute(
                """
                INSERT INTO rate_limit_counters (key, hits, expires_at)
                VALUES (%s, 1, %s)
                ON CONFLICT (key) DO UPDATE SET
                hits = CASE WHEN rate_limit_counters.expires_at <= %s THEN 1 ELSE rate_limit_counters.hits + 1 END,
                expires_at = CASE WHEN rate_limit_counters.expires_at <= %s THEN EXCLUDED.expires_at ELSE rate_limit_counters.expires_at END
                RETURNING hits, expires_at;
                """,
                [full_key, expiry, now, now],
            )
            row = cursor.fetchone()
            if not row:
                return False, 0, now

            hits, reset_time = row
            return (hits <= self.max_hits), hits, reset_time

    def reset(self, key: str):
        """Manually resets the counter for a key."""
        full_key = f"{self.prefix}:{key}"
        with connection.cursor() as cursor:
            cursor.execute("DELETE FROM rate_limit_counters WHERE key = %s", [full_key])
